File: verify.c
/*
 * Verification program for "A Three-Parameter Reduction of the Perfect Cuboid
 * Problem, and the Resolution of a Distinguished Family".
 *
 * Same computation as verify.py, fast enough for the x <= 7000 range quoted in
 * Section 6.  Uses __int128 because the radicand R exceeds 64 bits.
 *
 *   gcc -O2 -o verify verify.c -lm
 *   ./verify 7000
 *
 * Runtime: a few seconds at N=2000, about two minutes at N=7000.
 *
 * Expected output at N=7000 (final line):
 *   famA 719 (m1 square 719) | product-only 492 (m1 square 0)
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <math.h>

typedef __int128 i128;

/* quick non-residue filters: a square must be a square modulo each of these */
static int t64[64], t63[63], t65[65], t11[11];

static void init_tables(void)
{
    for (int i = 0; i < 64; i++) t64[i] = 0;
    for (int i = 0; i < 64; i++) t64[(i * i) % 64] = 1;
    for (int i = 0; i < 63; i++) t63[i] = 0;
    for (int i = 0; i < 63; i++) t63[(i * i) % 63] = 1;
    for (int i = 0; i < 65; i++) t65[i] = 0;
    for (int i = 0; i < 65; i++) t65[(i * i) % 65] = 1;
    for (int i = 0; i < 11; i++) t11[i] = 0;
    for (int i = 0; i < 11; i++) t11[(i * i) % 11] = 1;
}

static long long isqrt128(i128 n)
{
    long long r = (long long) sqrtl((long double) n);
    while ((i128) r * r > n) r--;
    while ((i128)(r + 1) * (r + 1) <= n) r++;
    return r;
}

static int is_square_ll(long long n)
{
    if (n < 0) return 0;
    long long r = isqrt128((i128) n);
    return r * r == n;
}

static long long gcd_ll(long long a, long long b)
{
    while (b) { long long t = a % b; a = b; b = t; }
    return a;
}

int main(int argc, char **argv)
{
    init_tables();
    long long N = (argc > 1) ? atoll(argv[1]) : 2000;

    long long famA = 0, prod = 0, qpos = 0, integral = 0, square = 0;
    long long m1sqA = 0, m1sqC = 0, next_report = 1000;

    for (long long x = 3; x <= N; x += 2) {
        i128 xx = (i128) x * x;
        for (long long y = 3; y < x; y += 2) {
            i128 yy = (i128) y * y, f1 = xx + yy;
            for (long long z = 1; z < y; z += 2) {
                i128 zz = (i128) z * z;
                i128 R = f1 * (xx - zz) * (yy - zz);

                if (!t64[(uint64_t) R & 63]) continue;
                uint64_t r = (uint64_t)(R % 45045);
                if (!t63[r % 63] || !t65[r % 65] || !t11[r % 11]) continue;

                long long k = isqrt128(R);
                if ((i128) k * k != R) continue;
                if (gcd_ll(gcd_ll(x, y), z) != 1) continue;

                i128 m = (i128) x * y * z, S = xx + yy - zz;
                long long h = gcd_ll((long long) m, k);
                long long m1 = (long long)(m / h);

                if ((i128) y * y == (i128) x * z) {          /* family A */
                    famA++;
                    if (is_square_ll(m1)) m1sqA++;
                    continue;
                }
                prod++;
                if (is_square_ll(m1)) m1sqC++;

                if (m <= k) continue;                        /* Q <= 0 */
                qpos++;

                long long d = m1 - k / h;
                if (S % d) continue;                         /* A^2 not integral */
                integral++;
                i128 A2 = (i128) m1 * (S / d);
                long long a2 = (long long) A2;
                if (is_square_ll(a2)) square++;
                printf("   integral A^2: (%lld,%lld,%lld)  m1=%lld  e=%lld  "
                       "A^2=%lld  square=%d\n",
                       x, y, z, m1, (long long)(S / d), a2, is_square_ll(a2));
            }
        }
        if (x >= next_report) {
            printf("x<=%6lld : famA %4lld (m1 sq %4lld) | product-only %4lld "
                   "(m1 sq %4lld) | Q>0 %4lld | d|S %2lld | A^2 square %lld\n",
                   x, famA, m1sqA, prod, m1sqC, qpos, integral, square);
            fflush(stdout);
            next_report += 1000;
        }
    }

    printf("\nFINAL N=%lld: famA %lld (m1 square %lld) | "
           "product-only %lld (m1 square %lld)\n",
           N, famA, m1sqA, prod, m1sqC);
    printf("            Q>0 %lld | d|S %lld | A^2 a perfect square %lld\n",
           qpos, integral, square);
    return 0;
}

File: verify.py
#!/usr/bin/env python3
"""
Verification script for "A Three-Parameter Reduction of the Perfect Cuboid
Problem, and the Resolution of a Distinguished Family".

Reproduces Table 1 and the counts quoted in Section 6.

Usage:   python3 verify.py 2000
Runtime: roughly 2 minutes at N=2000 on a typical laptop.
         Pure standard library; no packages required.

Expected output at N=2000:

    family A (y^2 = xz)                    199
    product-only (y^2 != xz)               158
      Q > 0                                 28
        d | S   (A^2 an integer)             2
          A^2 a perfect square               0
    sf(m1) >= 2x^2 among the 158            66
    m1 a perfect square: famA 199/199, product-only 0/158
"""

import sys
from math import isqrt, gcd


def is_square(n):
    if n < 0:
        return False
    r = isqrt(n)
    return r * r == n


def squarefree_kernel(n):
    """Product of primes dividing n to an odd power."""
    k = 1
    d = 2
    while d * d <= n:
        if n % d == 0:
            e = 0
            while n % d == 0:
                n //= d
                e += 1
            if e % 2:
                k *= d
        d += 1 if d == 2 else 2
    if n > 1:
        k *= n
    return k


def main(N):
    fam_a = prod_only = 0
    q_pos = integral = square = 0
    sf_fail = 0
    m1sq_a = m1sq_c = 0
    integral_cases = []

    # x > y > z > 0, all odd, gcd(x, y, z) = 1
    for x in range(3, N, 2):
        xx = x * x
        for y in range(3, x, 2):
            yy = y * y
            f1 = xx + yy
            for z in range(1, y, 2):
                if gcd(gcd(x, y), z) != 1:
                    continue
                zz = z * z
                R = f1 * (xx - zz) * (yy - zz)
                if not is_square(R):
                    continue

                k = isqrt(R)
                m = x * y * z
                S = xx + yy - zz
                h = gcd(m, k)
                m1 = m // h

                if y * y == x * z:                      # family A
                    fam_a += 1
                    if is_square(m1):
                        m1sq_a += 1
                    continue

                prod_only += 1
                if is_square(m1):
                    m1sq_c += 1
                if squarefree_kernel(m1) >= 2 * xx:
                    sf_fail += 1

                if m <= k:                              # Q <= 0, so A^2 <= 0
                    continue
                q_pos += 1

                d = m1 - k // h
                if S % d:                               # A^2 not an integer
                    continue
                integral += 1
                A2 = m1 * (S // d)
                assert A2 * (m - k) == m * S            # cross-check
                integral_cases.append((x, y, z, m1, S // d, A2))
                if is_square(A2):
                    square += 1

    print(f"N = {N}\n")
    print(f"  family A (y^2 = xz)                  {fam_a:6d}")
    print(f"  product-only (y^2 != xz)             {prod_only:6d}")
    print(f"    Q > 0                              {q_pos:6d}")
    print(f"      d | S   (A^2 an integer)         {integral:6d}")
    print(f"        A^2 a perfect square           {square:6d}")
    print()
    print(f"  sf(m1) >= 2x^2 among product-only    {sf_fail:6d}")
    print(f"  m1 a perfect square: famA {m1sq_a}/{fam_a}, "
          f"product-only {m1sq_c}/{prod_only}")
    print()
    for x, y, z, m1, e, A2 in integral_cases:
        print(f"  integral A^2: (x,y,z)=({x},{y},{z})  "
              f"m1={m1}  e={e}  A^2={A2}  square={is_square(A2)}")


if __name__ == "__main__":
    main(int(sys.argv[1]) if len(sys.argv) > 1 else 2000)
